pkgs/ffigen/lib/src/code_generator/objc_protocol.dart

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:ffigen/src/code_generator.dart';
+
+import 'binding_string.dart';
+import 'utils.dart';
+import 'writer.dart';
+
+class ObjCProtocol extends NoLookUpBinding with ObjCMethods {
+  final superProtos = <ObjCProtocol>[];
+  final String lookupName;
+  late final ObjCInternalGlobal _protocolPointer;
+
+  @override
+  final ObjCBuiltInFunctions builtInFunctions;
+
+  ObjCProtocol({
+    super.usr,
+    required String super.originalName,
+    String? name,
+    String? lookupName,
+    super.dartDoc,
+    required this.builtInFunctions,
+  })  : lookupName = lookupName ?? originalName,
+        super(name: name ?? originalName);
+
+  @override
+  BindingString toBindingString(Writer w) {
+    final protoMethod = ObjCBuiltInFunctions.protoMethod.gen(w);
+    final protoBuilder = ObjCBuiltInFunctions.protoBuilder.gen(w);
+    final objectBase = ObjCBuiltInFunctions.objectBase.gen(w);
+    final getSignature = ObjCBuiltInFunctions.getProtocolMethodSignature.gen(w);
+
+    final buildArgs = <String>[];
+    final buildImplementations = StringBuffer();
+    final methodFields = StringBuffer();
+
+    final uniqueNamer = UniqueNamer({name, 'pointer'});
+
+    for (final method in methods) {
+      final methodName = method.getDartMethodName(uniqueNamer);
+      final fieldName = methodName;
+      final argName = methodName;
+      final block = method.protocolBlock!;
+      final blockType = block.getDartType(w);
+
+      // The function type omits the first arg of the block, which is unused.
+      final func = FunctionType(returnType: block.returnType, parameters: [
+        for (int i = 1; i < block.argTypes.length; ++i)
+          Parameter(name: 'arg$i', type: block.argTypes[i]),
+      ]);
+      final funcType = func.getDartType(w, writeArgumentNames: false);
+
+      if (method.isOptional) {
+        buildArgs.add('$funcType? $argName');
+      } else {
+        buildArgs.add('required $funcType $argName');
+      }
+
+      final blockFirstArg = block.argTypes[0].getDartType(w);
+      final argsReceived = func.parameters
+          .map((p) => '${p.type.getDartType(w)} ${p.name}')
+          .join(', ');
+      final argsPassed = func.parameters.map((p) => p.name).join(', ');
+      final wrapper = '($blockFirstArg _, $argsReceived) => func($argsPassed)';
+
+      late final String listenerBuilder;
+      if (block.hasListener) {
+        listenerBuilder = '(Function func) => $blockType.listener($wrapper)';
+      } else {
+        final msg = '${method.originalName} cannot be a listener because '
+            'it has a non-void return type';
+        listenerBuilder = "(Function func) => throw UnsupportedError('$msg')";
```

Comment:
Why generate a method that throws if we know already at code gen time it's not going to work?

Maybe we should not generate a method at all?

Reply:
I was going to give an answer about how we have to construct a `ObjCProtocolMethod` to pass to the `ObjCProtocolBuilder`. There were more ergonomic solutions (with compile time type checking of the function) where I would generate a different method class per method, but Dart doesn't have nested classes so it would have been pretty messy. But then I realized I could at least split listenable vs non-listenable methods.

So methods that have a return value become `ObjCProtocolMethod`, and void methods become `ObjCProtocolListenableMethod`. So we can at least compile time check that you're not trying to implement a non-void method as a listener.